File: LAB_10/Command.java
import java.util.ArrayList;
import java.util.Stack;


interface Command1 {
    void execute();
    void undo();
}

class TextEditor {
    private final ArrayList<String> text = new ArrayList<>();

    public void addText(String newText) {
        text.add(newText);
        System.out.println("Added tekxt: " + newText);
    }

    public void removeText(int length) {
        String removedText = text.removeLast();
        System.out.println("Removed text: " + removedText);
    }

    public String getText() {
        return String.join("", text);
    }
}

class AddTextCommand implements Command1 {
    private final TextEditor editor;
    private final String text;

    public AddTextCommand(TextEditor editor, String text) {
        this.editor = editor;
        this.text = text;
    }

    @Override
    public void execute() {
        editor.addText(text);
    }

    @Override
    public void undo() {
        editor.removeText(text.length());
    }
}

class CommandManager {
    private final Stack<Command1> commandHistory = new Stack<>();
    private final Stack<Command1> redoStack = new Stack<>();

    public void executeCommand(Command1 command) {
        command.execute();
        commandHistory.push(command);
        redoStack.clear();
    }

    public void undo() {
        if (!commandHistory.isEmpty()) {
            Command1 command = commandHistory.pop();
            command.undo();
            redoStack.push(command);
        } else {
            System.out.println("No commands to undo!");
        }
    }

    public void redo() {
        if (!redoStack.isEmpty()) {
            Command1 command = redoStack.pop();
            command.execute();
            commandHistory.push(command);
        } else {
            System.out.println("No commands to redo!");
        }
    }
}

public class Command {
    public static void main(String[] args) {
        TextEditor editor = new TextEditor();

        CommandManager manager = new CommandManager();

        Command1 addHello = new AddTextCommand(editor, "Hello ");
        Command1 addWorld = new AddTextCommand(editor, "World!");

        manager.executeCommand(addHello);
        manager.executeCommand(addWorld);
        System.out.println("Current text: " + editor.getText());

        manager.undo();
        System.out.println("After undo: " + editor.getText());

        manager.redo();
        System.out.println("After redo: " + editor.getText());
    }
}

File: LAB_10/Command.cpp
#include <iostream>
#include <stack>
#include <vector>
#include <string>


class Command1 {
    public:
        virtual void execute()=0;
        virtual void undo()=0;
};

class TextEditor {
    private:
        std::vector<std::string>* text = new std::vector<std::string>();

    public:
        void addText(std::string newText) {
            text->push_back(newText);
            std::cout << "Added text: " << newText << std::endl;
    }

        void removeText(int length) {
            std::string removedText = text->back();
            text->pop_back();
            std::cout << "Removed text: " << removedText << std::endl;
        }

        std::string getText() {
            std::string result = "";
            for (std::string s : *text) {
                result += s;
            }
            return result;
        }
};

class AddTextCommand : public Command1 {
    private:
        TextEditor* editor;
        std::string text;

    public:
        AddTextCommand(TextEditor* editor, std::string text) {
            this->editor = editor;
            this->text = text;
        }

        void execute() override {
            editor->addText(text);
        }

        void undo() override {
            editor->removeText(text.length());
        }
};

class CommandManager {
    private:
        std::stack<Command1*>* commandHistory = new std::stack<Command1*>();
        std::stack<Command1*>* redoHistory = new std::stack<Command1*>();

    public:
        void executeCommand(Command1* command) {
            command->execute();
            commandHistory->push(command);
            while (!redoHistory->empty()) {
                redoHistory->pop();
            }
        }

        void undo() {
            if (!commandHistory->empty()) {
                Command1* command = commandHistory->top();
                commandHistory->pop();
                command->undo();
                redoHistory->push(command);
            } else {
                std::cout << "No commands to undo!" << std::endl;
            }
    }

        void redo() {
            if (!redoHistory->empty()) {
                Command1* command = redoHistory->top();
                redoHistory->pop();
                command->execute();
                commandHistory->push(command);
            } else {
                std::cout << "No commands to redo!" << std::endl;
            }
        }
};


int main() {
    TextEditor* editor = new TextEditor();

    CommandManager* manager = new CommandManager();

    Command1* addHello = new AddTextCommand(editor, "Hello ");
    Command1* addWorld = new AddTextCommand(editor, "World!");

    manager->executeCommand(addHello);
    manager->executeCommand(addWorld);
    std::cout << "Current text: " << editor->getText() << std::endl;

    manager->undo();
    std::cout << "After undo: " << editor->getText() << std::endl;

    manager->redo();
    std::cout << "After redo: " << editor->getText() << std::endl;

    delete editor;
    delete manager;
    delete addHello;
    delete addWorld;

    return 0;
}